ndcube/asdf/converters/ndcube_converter.py

```diff
@@ -0,0 +1,30 @@
+from asdf.extension import Converter
+import numpy as np
+
+
+class NDCubeConverter(Converter):
+    tags = ["tag:sunpy.org:ndcube/ndcube/NDCube-*"]
+
+    @property
+    def types(self):
+        from ndcube.ndcube import NDCube
+        return [NDCube]
+
+    def select_tag(self, obj, tags, ctx):
+        # Sort the tags in reverse alphabetical order and pick the first (i.e.
+        # the one with the highest version). This assumes that all the tags for
+        # this converter are named the same other than the version number.
+        tags = list(sorted(tags, reverse=True))
+        return tags[0]
+
+    def from_yaml_tree(self, node, tag, ctx):
+        from ndcube import ndcube
+        data = np.asanyarray(node["data"])
```

Comment:
Was `asanyarray` needed because `NDCube` won't accept `NDArrayType`?

Reply:
I have used `np.asanyarray` to ensure that the data attribute maintains the `numpy.ndarray` type after deserialization. So, when an NDCube ASDF file is deserialized, the `.data` attribute retains its original `numpy.ndarray` type instead of an `asdf.tags.core.ndarray.NDArrayType` type. This would ensures consistency in the data type.

Reply:
Thanks for the detailed response. I ask because `NDArrayType` is "lazy", it will only read the array data from the file when it's needed (not when the `NDArrayType` is created). If we have an ASDF file "foo.asdf" with an array under key "arr"
```python
>> af = asdf.open("foo.asdf")
>> print(af["arr"])  # print and some other operations (like shape) won't load the data
<array (unloaded) shape: [100, 100] dtype: float64>
```
Passing the `NDArrayType` through `asanyarray` will convert it to a `ndarray` (as you noted) and trigger the data to be loaded. I had a quick look and it appears that `NDCube` will accept the `NDArrayType`. However something in the constructor is triggering the `NDArrayType` to load the array data (this can happen even for seemingly innocuous calls like this call to [hasattr](https://github.com/sunpy/ndcube/blob/ce4d0b13705fcb8128a87402a23ac0378e7c58da/ndcube/ndcube.py#L379)).
Since the conversion to `ndarray` doesn't seem to copy the data (I tested this by providing the result and input to [shares_memory](https://numpy.org/doc/stable/reference/generated/numpy.shares_memory.html#numpy-shares-memory)) and the `NDCube` constructor triggers the data loading it seems like a good idea to include this `asanyarray` so that `NDCube` doesn't have to handle the `NDArrayType`. Does that sound good to you?

Reply:
Yes, it makes sense, the loaded data is getting passed to the `NDCube` and it doesn't need to handle the loading behaviour of `NDArrayType`